org-eclipse-che-jdt-ls-extension/jdt.ls.extension.core/src/main/java/org/eclipse/che/jdt/ls/extension/core/internal/JavaModelEventProvider.java

```diff
@@ -0,0 +1,106 @@
+/*
+ * Copyright (c) 2012-2018 Red Hat, Inc.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.che.jdt.ls.extension.core.internal;
+
+import static org.eclipse.che.jdt.ls.extension.api.Commands.CLIENT_UPDATE_PROJECTS_CLASSPATH;
+
+import java.util.HashSet;
+import java.util.Set;
+import org.eclipse.core.resources.IProject;
+import org.eclipse.jdt.core.ElementChangedEvent;
+import org.eclipse.jdt.core.IElementChangedListener;
+import org.eclipse.jdt.core.IJavaElementDelta;
+import org.eclipse.jdt.core.JavaCore;
+import org.eclipse.jdt.ls.core.internal.JavaLanguageServerPlugin;
+import org.eclipse.jdt.ls.core.internal.ResourceUtils;
+import org.eclipse.jdt.ls.core.internal.handlers.JDTLanguageServer;
+
+/**
+ * Provides Java Model Element Change events to the clients
+ *
+ * @author Victor Rubezhny
+ */
+@SuppressWarnings("restriction")
+public class JavaModelEventProvider implements IElementChangedListener {
+
+  private static final int CLASSPATH_CHANGED_MASK =
+      IJavaElementDelta.F_ADDED_TO_CLASSPATH
+          | IJavaElementDelta.F_CLASSPATH_CHANGED
+          | IJavaElementDelta.F_REORDER
+          | IJavaElementDelta.F_REMOVED_FROM_CLASSPATH
+          | IJavaElementDelta.F_RESOLVED_CLASSPATH_CHANGED
+          | IJavaElementDelta.F_ARCHIVE_CONTENT_CHANGED;
+
+  private static JavaModelEventProvider INSTANCE;
+
+  private boolean isListening = false;
+
+  private JavaModelEventProvider() {}
+
+  public static JavaModelEventProvider getInstance() {
+    if (INSTANCE == null) {
+      INSTANCE = new JavaModelEventProvider();
+    }
+    return INSTANCE;
+  }
+
+  public void start() {
+    if (!this.isListening) {
+      this.isListening = true;
+      JavaCore.addElementChangedListener(this);
+    }
+  }
+
+  public void stop() {
+    if (this.isListening) {
+      this.isListening = false;
+      JavaCore.removeElementChangedListener(this);
+    }
+  }
+
+  @Override
+  public void elementChanged(ElementChangedEvent event) {
+    Set<IProject> projects = getAffectedProjects(event.getDelta(), new HashSet<IProject>());
+    if (projects.isEmpty()) {
+      return;
+    }
+
+    try {
+      Set<String> projectUris = new HashSet<String>();
+      for (IProject project : projects) {
+        projectUris.add(ResourceUtils.fixURI(project.getRawLocationURI()));
```

Comment:
Why "getRawLocationURI()"?

Reply:
Ocasionously was copied from somewhere... Probably getLocationURI() should be used instead, But now I'n not sure if I do need to have it in URI form as the only relative path is actually used.